### ISSUE TITLE
[BUG]: `Slicing with [::-1]` becomes `[0: 9223372036854775807 : -1]` and has a negative length

### ISSUE THREAD

**Issue by mikowals** [modularml/mojo#1944]
### Bug description

I expect slicing with `[::-1]` to produce something that can be bounds checked and used to reverse an entire sequence. The correct length of `Slice(0, 9223372036854775807, -1)` should be 0 and not -9223372036854775807.

Or the compiler could force start and end to be provided when step is negative.

The new `Slice.__init__[T0: AnyRegType, T1: AnyRegType, T2: AnyRegType](start: T0, end: T1, step: T2) -> Self` allows working around this with custom slice. But this is a sharp edge when using the default Slice provided by sugared syntax.

### Steps to reproduce

This code:

```mojo
@value
struct Array:
    var size: Int
 
    fn __getitem__(self, s: Slice) -> Int:
        print("start:", s.start, "end:", s.end, "step:", s.step)
        print("slice length: ", len(s))
        print("s[0]:", s[0], "s[1]:", s[1])
        return 0
 
    fn main():
        var y = Array(1)
        print("y[::-1]")
        _ = y[::-1]
```

produces:

```console
y[::-1]
start: 0 end: 9223372036854775807 step: -1
slice length:  -9223372036854775807
s[0]: 0 s[1]: -1
```

### System information

```shell
MacOS using Mojo 24.1.0
```

**Comment by LJ-9801** [comment]
can we default the `end` to -1 instead of `_int_max_value()? at least now the end always points to the last element instead of this unsafe memory access.

**Comment by mikowals** [comment]
I think ultimately it will need to be done as in Python with both start and end able to hold a `None` to represent the users intent.  Using `-1` as any default runs into conflict with the Python standard where negative indices count back from the end.

**Comment by mikowals** [comment]
This is fixed by #2495.
